File: 200910/Week08/ChangeMaking.cpp
// Oliver Kullmann, 16.3.2010 (Swansea)

/*!
  \file ChangeMaking.cpp
  \brief Application for exploring the change-making algorithm

  Use by
  \verbatim
ChangeMaking N d1 ... dn
  \endverbatim
  where N is the amount to be returned, n is the number of coins and
  d1, ..., dn are the coin values.
*/

#include <iostream>
#include <sstream>
#include <vector>
#include <cassert>
#include <string>
#include <iomanip>

#include "ChangeMaking.hpp"

namespace {

  enum { errcode_parameter = 1, errcode_parameter_reading = 2,
         errcode_parameter2_reading = 3, errcode_nonpositive = 4 };

  const std::string program = "ChangeMaking";
  const std::string err = "ERROR[" + program + "]: ";

  using DynamicProgramming::size_type;
  using DynamicProgramming::vector_type;
  using DynamicProgramming::matrix_type;

  inline void size_out(const size_type s) {
    if (s == DynamicProgramming::infinity)
      std::cout << "inf";
    else
      std::cout << s;
  }

}

int main(const int argc, const char* const argv[]) {

  if (argc < 2) {
    std::cerr << err << "At least one parameter is needed, the amount N to be returned.\n";
    return errcode_parameter;
  }

  size_type dummy;
  {
    std::stringstream parameters;
    parameters << argv[1];
    parameters >> dummy;
    if (not parameters) {
      std::cerr << err << "Error when reading the amount N (which must be a non-negative integer).\n";
      return errcode_parameter_reading;
    }
  }
  const size_type N = dummy;
  const size_type n = argc-2;
  if (n == 0) {
    if (N == 0)
      std::cout << "No coins needed.\n";
    else
      std::cout << "Without coins no solution.\n";
    return 0;
  }
  vector_type d;
  for (size_type i = 2; i < (size_type) argc; ++i) {
    std::stringstream parameters;
    parameters << argv[i];
    parameters >> dummy;
    if (not parameters) {
      std::cerr << err << "Error when reading coin " << i-1 << " (which must be a non-negative integer).\n";
      return errcode_parameter2_reading;
    }
    if (dummy == 0) {
      std::cerr << err << "Coin " << i-1 << " is zero.\n";
      return errcode_nonpositive;
    }
    d.push_back(dummy);
  }
  assert(d.size() == n);

  const matrix_type c = DynamicProgramming::making_change(N, d);
  std::cout << "Optimal number of coins needed: ";
  size_out(c.back()[N]);
  std::cout << "\n";

  if (c.back()[N] != DynamicProgramming::infinity) {
    std::cout << "Solution: ";
    const vector_type a = DynamicProgramming::pay_out(c,d);
    assert(a.size() == n);
    for (size_type i = 0; i < n; ++i) {
      std::cout << a[i] << "*" << d[i];
      if (i+1 < n) std::cout << " + ";
    }
    std::cout << " = " << N << "\n";
  }

  std::cout << "\nOutput of array c ? Enter \"y\" or \"n\": ";
  std::string question;
  std::cin >> question;
  if (question == "y")
    for (size_type i = 0; i < n; ++i) {
      std::cout << i+1 << ": ";
      for (size_type j = 0; j <= N; ++j) {
        std::cout << std::setw(3);
        size_out(c[i][j]);
        std::cout << " ";
      }
      std::cout << "\n";
    }

}

File: 201112/Week09/ChangeMaking.java
// Arnold Beckmann, 20.03.2011 (Swansea)
// Oliver Kullmann, 30.11.2011 (Swansea)

/*
  File:  201011/Week09/ChangeMaking.java
  Descr: Solving the change-making problem via dynamic programming
*/

class ChangeMaking {

    public final static int infinity = Integer.MAX_VALUE;

    public static int[][] making_change(final int N, final int[] d) {
      final int n = d.length;
      int[][] c = new int[n + 1][N + 1];
      if (n == 0) return c;
      assert(N < infinity);

      for (int j = 1; j <= N; ++j) c[0][j] = infinity;
      for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= N; ++j)
          if (j < d[i-1] || c[i][j-d[i-1]] == infinity)
            c[i][j] = c[i-1][j];
          else
            c[i][j] = Math.min(c[i-1][j], 1+c[i][j-d[i-1]]);
      return c;
    }

    public static int[] pay_out(final int[][] c, final int[] d) {
      final int n = d.length;
      final int[] a = new int[n];
      if (n == 0) return a;

      final int N = c[0].length - 1;
      int i = n, j = N;
      while (j > 0)
	  if (i == 0) { j = 0; }
	  else if (c[i][j] == c[i-1][j]) --i;
	  else {
	    ++a[i-1];
	    j -= d[i-1];
	  }
      return a;
    }

    public static int rec_making_change(final int N, final int[] d) {
	final int n = d.length;
	final int nc = rec_making_change(d, n, N);
	return nc;
    }

    private static int rec_making_change(final int[] d,
					final int i, final int j) {
	if (j <= 0) return 0;
	if (i <= 0) return infinity;
	final int nc1 = rec_making_change(d, i-1, j);
	if (j < d[i-1])  return nc1;
	final int nc2 = rec_making_change(d, i, j-d[i-1]);
	return Math.min(nc1, 1+nc2);
    }


}		


